File: python/114_FlattenBinaryTreeToLinkedList.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def flatten(self, root: TreeNode) -> None:
        """
        Do not return anything, modify root in-place instead.
        """
        if not root:
            return None
        if not root.left and not root.right:
            return root
        
        left, right = self.flatten(root.left), self.flatten(root.right)
        
        if left:
            left.right = root.right
            root.left, root.right = None, root.left
        
        return right if right else left

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def flatten(self, root: TreeNode) -> None:
        """
        Do not return anything, modify root in-place instead.
        """
        if not root:
            return None
        
        left = self.flatten(root.left) if root.left else None
        right = self.flatten(root.right) if root.right else None
        if left:
            root.left, root.right = None, left
            curr = left
            while curr.right:
                curr = curr.right
            curr.right = right
            
        return root

File: cpp/114_FlattenBinaryTreeToLinkedList.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    void flatten(TreeNode* root) {
        flattenTree(root);
    }
    
    TreeNode* flattenTree(TreeNode* node) {
        if (node == nullptr) {
            return nullptr;
        } else if (node->right == nullptr and node->left == nullptr) {
            return node;
        }
        
        TreeNode* left = flattenTree(node->left);
        TreeNode* right = flattenTree(node->right);
        
        if (left != nullptr) {
            left->right = node->right;
            node->right = node->left;
            node->left = nullptr;
        }
        
        return right == nullptr ? left : right;
            
    }
};
